### DISCUSSION TITLE
How to pass a list of `ContentItem` to Liquid so it can be rendered?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#14085] [Q&A] [answered]
I am trying to pass a list of `ContentItem` to Liquid so I can use Liquid to loop throw the records.

Here is how I am processing a Liquid message

```csharp
    private async Task<string> GetContentAsync(string message, IUser user, ContentItem contentItem, IDictionary<string, ContentItem> events)
    {
        var result = await _liquidTemplateManager.RenderHtmlContentAsync(message, _htmlEncoder, null,
            new Dictionary<string, FluidValue>()
            {
                ["User"] = new ObjectValue(user),
                ["ContentItem"] = new ObjectValue(contentItem),
                ["ContentEvents"] = new DictionaryValue(new ObjectDictionaryFluidIndexable<ContentItem>(events, _templateOptions)),
            });

        using var writer = new StringWriter();
        result.WriteTo(writer, _htmlEncoder);

        return writer.ToString();
    }
```

Then here is my Liquid syntax

```
{% assign siteRoot = '~/' | absolute_url %}

Here is a list of the coming up events:

<table>
{% for contentEvent in ContentEvents %}
<tr>
    <td><a href="{{ contentEvent | display_url | prepend: siteRoot}}">{{ contentEvent | display_text }}</a></td>
</tr>
{% endfor %}
</table>
```

The loop seems to work. But, `{{ contentEvent | display_text }}` does not seems to return anything. Here is the output

```
<table>
<tr>
    <td><a href="https://localhost/ed12/"></a></td>
</tr>
<tr>
    <td><a href="https://localhost/ed12/"></a></td>
</tr>
</table>
```

@jtkech any idea what am I missing here? I am guessing `DictionaryValue(new ObjectDictionaryFluidIndexable<ContentItem>(events, _templateOptions))` is wrong. But, not sure what is the right way to go about it.

**Comment by jtkech** [comment, marked as answer]
I don't know, I assume that you retrieve `_templateOptions` through `IOptions<>`.

Maybe use `IDictionary<string, object>` or `IDictionary<string, FluidValue>`.

And/or use `FluidValue.Create(ContentEvents, _templateOptions)`.

Hmm, I don't think that a `MemberAccessStrategy` is needed for the dictionary itself.

**Comment by MikeAlhayek** [comment]
> I don't know, I assume that you retrieve _templateOptions through IOptions<>.

Yes I am.

> Maybe use IDictionary<string, object> or IDictionary<string, FluidValue>.

Tried both with no luck.

But using the following finally worked!
```
private async Task<string> GetContentAsync(string message, IUser user, ContentItem contentItem, IEnumerable<ContentItem> contentItems)
{    
    var result = await _liquidTemplateManager.RenderHtmlContentAsync(message, _htmlEncoder, null,
        new Dictionary<string, FluidValue>()
        {
            ["User"] = new ObjectValue(user),
            ["ContentItem"] = new ObjectValue(contentItem),
            ["ContentEvents"] = FluidValue.Create(contentItems, _templateOptions),
        });

    using var writer = new StringWriter();
    result.WriteTo(writer, _htmlEncoder);

    return writer.ToString();
}
```
Thank you for your feedback!

**Comment by jtkech** [comment]
Yes because you were looping through a dictionary returning KeyValuePairs
